Mathlib/LinearAlgebra/FreeModule/Int.lean

```diff
@@ -0,0 +1,191 @@
+              rintro j - hj
+              have hinj := f.injective.ne hj
+              rw [h.choose_spec] at hinj
+              simp [hinj.symm]
+          _ = c h.choose * a h.choose := by simp [h.choose_spec]
+      · convert dvd_refl (0 : R)
+        convert Finset.sum_const_zero with j
+        rw [not_exists] at h
+        specialize h j
+        rw [eq_comm] at h
+        simp [h]
+    · refine ⟨fun j ↦ (h (f j)).choose, ?_⟩
+      ext i
+      simp only [EmbeddingLike.apply_eq_iff_eq, exists_eq, ↓reduceDIte, Classical.choose_eq,
+        Finset.sum_apply, Pi.smul_apply, Pi.single_apply, smul_ite, smul_zero]
+      rw [eq_comm]
+      by_cases hj : ∃ j, f j = i
+      · calc ∑ x : Fin n, _ =
+            if i = f hj.choose then (h (f hj.choose)).choose * a hj.choose else 0 := by
+              convert Finset.sum_eq_single (β := R) hj.choose ?_ ?_
+              · simp [hj]
+              · rintro j - h
+                have hinj := f.injective.ne h
+                rw [hj.choose_spec] at hinj
+                simp [hinj.symm]
+              · simp
+          _ = g i := by
+              simp only [hj.choose_spec, ↓reduceIte]
+              rw [mul_comm]
+              conv_rhs =>
+                rw [← hj.choose_spec, (h (f hj.choose)).choose_spec]
+              simp only [EmbeddingLike.apply_eq_iff_eq, exists_eq, ↓reduceDIte, Classical.choose_eq]
+              convert rfl
+              · exact hj.choose_spec
+              · simp [hj]
+      · convert Finset.sum_const_zero with x
+        · rw [not_exists] at hj
+          specialize hj x
+          rw [eq_comm] at hj
+          simp [hj]
+        · rw [← zero_dvd_iff]
+          convert h i
+          simp [hj]
+  rw [hN', AddSubgroup.index_pi]
+  split_ifs with h
+  · have hb : Function.Bijective f :=
+      (Nat.bijective_iff_injective_and_card _).2 ⟨f.injective, by simp [h]⟩
+    let fe : Fin n ≃ ι := Equiv.ofBijective f hb
+    convert Finset.prod_map _ fe.toEmbedding _ with - - i -
+    · simp
+    · have he : ∃ j, f j = fe.toEmbedding i := ⟨i, rfl⟩
+      rw [dif_pos he]
+      convert rfl
+      exact f.injective he.choose_spec
+  · simp only [Finset.prod_eq_zero_iff, Finset.mem_univ, true_and]
+    have hlt : Fintype.card (Fin n) < Fintype.card ι :=
+      Ne.lt_of_le (by simpa using h) (Fintype.card_le_of_embedding f)
+    have hi : (Finset.image f Finset.univ).card < Fintype.card ι :=
+      Finset.card_image_le.trans_lt (by simpa using hlt)
+    have hu : ¬(Finset.image f Finset.univ) = Finset.univ := by
+      rw [← Finset.card_eq_iff_eq_univ]
+      exact hi.ne
+    simp only [Finset.eq_univ_iff_forall, Finset.mem_image, Finset.mem_univ, true_and, not_forall,
+      not_exists] at hu
+    rcases hu with ⟨i, hi⟩
+    rw [← not_exists] at hi
+    refine ⟨i, ?_⟩
+    simp only [hi, ↓reduceDIte, Set.singleton_zero, Ideal.span_zero, Submodule.bot_toAddSubgroup,
+      AddSubgroup.index_bot, Nat.card_eq_zero_of_infinite]
```

Comment:
This proof is a bit long, do you think it is possible to split out a sublemma?

Reply:
It may not have made things shorter, but I've split things up further (a lemma giving the index as a product not depending on `Infinite` separate from the simpler expression assuming `Infinite`).